src/Symfony/Component/Mime/Email.php

```diff
@@ -480,6 +480,12 @@ private function prepareParts(): ?array
         foreach ($this->attachments as $attachment) {
             foreach ($names as $name) {
                 if (isset($attachment['part'])) {
+                    if ('inline' === $attachment['part']->getPreparedHeaders()->getHeaderBody('Content-Disposition')) {
```

Comment:
this should be done before the loop over names (to do it even when there is no names being found for `<img>` tags) and it needs to replace the `cid` references in the HTML too. Otherwise, the fix is incomplete.

and as said in my comment, we also need a fix for attachments using `inline: true` without a part already.

Reply:
@stof maybe I'm missing something, but aren't inline parts handled by this already?

https://github.com/symfony/symfony/blob/7c531f575e7e14636378787271c29b094205e147/src/Symfony/Component/Mime/Email.php#L485-L488

Or do you mean that someone calls `embed` but the regex doesn't find the `$name` in the HTML body so that it drops out of the loop here?

https://github.com/symfony/symfony/blob/7c531f575e7e14636378787271c29b094205e147/src/Symfony/Component/Mime/Email.php#L479-L481

Because in that case (at least in my uninformed opinion) it seems correct to not make it an inline attachment, since it's actually not embedded anywhere.

Reply:
@flack no. this handles attachments that are added with a name that matches a name detected in the HTML. It does not handle the case of attachments that are added as `inline: true` explicitly through `->embed()`.

And `$names` is not detecting all usages of CID in the HTML (that's precisely the issue you are marking as closed by this PR, which is *not* solved by your PR)

Reply:
@stof well, I only said this fixes the issue _in the second comment_ of #42921, i.e. the allows me to build a workaround for the original problem in that ticket. It was @fabpot who added "closes #42921" to the description here, don't blame me for that :-)

But I still don't get it: Attachments that are added as `inline: true` explicitly through `->embed()` cannot have `$attachment['part']` set:

https://github.com/symfony/symfony/blob/7c531f575e7e14636378787271c29b094205e147/src/Symfony/Component/Mime/Email.php#L353-L358

So the handling of those doesn't change in this PR. Of course, if they were ignored before (like in the OP of #42921), they are still ignored. But that seems like a separate issue to me. This PR fixes a bug where you can't add an instantiated `DataPart` as inline. The other ticket is about the name detection logic being too narrow.

Reply:
> So the handling of those doesn't change in this PR.

and that's precisely what I'm asking to change, by actually fixing the reported issue instead of fixing only the workaround